system/mp/mpcomm_mpi.cpp

```diff
@@ -0,0 +1,302 @@
+//#include "jlib.hpp"
+#include <limits.h>
+
+#include <unistd.h>
+
+#include "mpcomm.hpp"
+#include "mpbuff.hpp"
+#include "mputil.hpp"
+#include "mplog.hpp"
+
+#include "mpi_wrapper.hpp"
+
+#ifdef _MSC_VER
+#pragma warning (disable : 4355)
+#endif
+
+//#define _TRACE
+//#define _FULLTRACE
+
+//#define _TRACEMPSERVERNOTIFYCLOSED
+#define _TRACEORPHANS
+
+
+#define REFUSE_STALE_CONNECTION
+
+
+#define MP_PROTOCOL_VERSION    0x102                   
+#define MP_PROTOCOL_VERSIONV6   0x202                   // extended for IPV6
+
+// These should really be configurable
+#define CANCELTIMEOUT       1000             // 1 sec
+#define CONNECT_TIMEOUT          (5*60*1000) // 5 mins
+#define CONNECT_READ_TIMEOUT     (10*1000)   // 10 seconds. NB: used by connect readtms loop (see loopCnt)
+#define CONNECT_TIMEOUT_INTERVAL 1000        // 1 sec
+#define CONNECT_RETRYCOUNT       180         // Overall max connect time is = CONNECT_RETRYCOUNT * CONNECT_READ_TIMEOUT
+#define CONNECT_TIMEOUT_MINSLEEP 2000        // random range: CONNECT_TIMEOUT_MINSLEEP to CONNECT_TIMEOUT_MAXSLEEP milliseconds
+#define CONNECT_TIMEOUT_MAXSLEEP 5000
+
+#define CONFIRM_TIMEOUT          (90*1000) // 1.5 mins
+#define CONFIRM_TIMEOUT_INTERVAL 5000 // 5 secs
+#define TRACESLOW_THRESHOLD      1000 // 1 sec
+
+#define VERIFY_DELAY            (1*60*1000)  // 1 Minute
+#define VERIFY_TIMEOUT          (1*60*1000)  // 1 Minute
+
+#define DIGIT1 U64C(0x10000000000) // (256ULL*256ULL*256ULL*65536ULL)
+#define DIGIT2 U64C(0x100000000)   // (256ULL*256ULL*65536ULL)
+#define DIGIT3 U64C(0x1000000)     // (256ULL*65536ULL)
+#define DIGIT4 U64C(0x10000)       // (65536ULL)
+
+#define _TRACING
+
+class NodeCommunicator: public ICommunicator, public CInterface{
+    NodeGroup *group;
+    bool outer;
+    rank_t myrank;
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    bool send(CMessageBuffer &mbuf, rank_t dstrank, mptag_t tag, unsigned timeout){ 
+        assertex(dstrank!=RANK_NULL);
+        CTimeMon tm(timeout);
+        rank_t myrank = group->rank();
+        rank_t startrank = dstrank;
+        rank_t endrank;
+        if (dstrank==RANK_ALL || dstrank==RANK_ALL_OTHER) {
+            startrank = 0;
+            endrank = group->ordinality()-1;
+        } else if (dstrank==RANK_RANDOM) {
+            if (group->ordinality()>1) {
+                do {
+                    startrank = getRandom()%group->ordinality();
+                } while (startrank==myrank);
+            } else {
+                assertex(myrank!=0);
+                startrank = 0;
+            }
+            endrank = startrank;
+        } else {
+            endrank = startrank;
+        }
+        for (;startrank<=endrank;startrank++) {
+            if ((startrank!=RANK_ALL_OTHER) || (startrank!=myrank)) {
+                unsigned remaining;
+                if (tm.timedout(&remaining))
+                    return false;
+                hpcc_mpi::CommRequest req = hpcc_mpi::sendData(startrank, tag, mbuf, *group, true);
```

Comment:
This is where how the design of the MP project clashes with how MPI work. For example all communication would happen within a group. A node/process in one IGroup cannot talk to a node/process in another IGroup. Therefore it makes sense to have all the communication functions such as send/recv/reply/cancel etc. inside IGroup rather than having it wrapped around a ICommunicator. 

I wanted to wrap all the MPI framework functions because it is the bottom layer in the communication hierarchy of the MP project. If there's something wrong with or requires improvements or changes in how MPI framework is used, we can change that layer independent of application logic expected of the ICommunicator or IGroup. Hence why I introduced mpi_wrapper.hpp/cpp. Everything related to MPI framework happens there. Let me know your thoughts on it.

Reply:
For our purposes a IGroup is only a logical definition of a group of 'nodes' and contains functions that deal with sets/ranks only.
NB: groups are divisble/mergable and published to Dali - they don't imply communication, just a definition.
So introducing in an implementation of IGroup which contains networking comm. changing it's scope beyond what it is intended for.

Whereas an ICommunicator is essentially a group of comms channels based on the logical group. An ICommunicator's implementation deals with all comms between ranks in a group. NB: it is dealing in general with a group of CMPChannel's at this point, it is only using the IGroup for positioning not comms, e.g. rank.

I see the IGroup/INode remaining as they are, such that with the same IGroup instance I can say:
```
IGroup *anIGroup = <from somewhere>;
ICommunicator *mpComm = createCommunicator(anIGroup);
ICommunicator *mpiComm = createCommunicator(anIGroup);
..
```